### PR TITLE
#167165061 #167165069 CRUD articles and create tags for articles

### PR DESCRIPTION
**Pull request by tolulope-od**
### What does this PR do?

Adds functionality to create articles and add tags to an article
Modifies database model for articles to add new data types
Modifies server response method
Adds validation when creating an article
Creates a route to get articles by tag
Creates a route to get an article by author and slug
Creates a route for an author to get their own articles
Adds more tags support up to 10
Change server success response format for profiles

#### Description of Task to be completed?
Users should be able to Create, Read, Update, Delete articles (WIP)
Create articles - `POST /api/v1/articles`
Get all published articles - `GET -> /api/v1/articles`
Get articles by a tag - `GET /api/v1/articles?tag=`
Get articles by slug - `GET /api/v1/:email/:slug`
Get all published articles by an author - `GET /api/v1/:email`
Get all articles by an author including drafts (for author) - `GET /api/v1/users/articles`

#### How should this be manually tested?
After cloning this repo, `cd` into it and do the following:

```bash
# fetch this branch
$ git fetch origin  ft-user-create-article-and-tags-167165061-167165069

# install dependencies
$ npm install

# run unit tests
$ npm test

# start the development server
$ npm run server
```
Using postman, log in or register an account and hit the post route to create an article (user must be verified via the verification route in #28 

add the following format in the request body:

```JSON
{
 "title": "How to live a fulfilling life",
  "description": "Do what you love",
  "body": "Lorem Ipsum",
  "tags": ["life", "advice", "career", "self-care"],
  "status": "published"
}
```

#### Any background context you want to provide?
N/a

#### What are the relevant pivotal tracker stories?
[#167165061](https://www.pivotaltracker.com/story/show/167165061) 
[#167165069](https://www.pivotaltracker.com/story/show/167165069)

#### Screenshots (if appropriate)

*Created article successfully*
Screenshot 2019-08-01 at 08 16 46

*Get all published articles by an author*
Screenshot 2019-08-01 at 08 16 46

*Get articles by a tag*

Screenshot 2019-08-01 at 08 19 06

### CODE REVIEW

**Comment by VictorMaria** [comment]
Well done Tolu, your tests are comprehensive. You may try rebasing you branch due to the merge conflicts. Good job.

**Comment by tolulope-od** [comment]
> Well done Tolu, your tests are comprehensive. You may try rebasing you branch due to the merge conflicts. Good job.

Thanks Victor, I've fixed rebased the branch

**Comment by tolulope-od** [comment]
I want to apologize for having so many diffs on one branch, I tried to break things down into commits to reduce the number of diffs, but the task was quite huge and required a lot of changes and additions.

**Review by Tunji17 — APPROVED** [review]
Great job @tolulope-od , LGTM

File: __tests__/articleRoutes.test.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  });
+
+  it('should throw an error when the tags of an article more than the allowed amount', (done) => {
+    chai.request(app)
+      .post(`${API_PREFIX}`).send(mockArticles[2]).set('Authorization', `Bearer ${validUserToken}`)
+      .end((err, res) => {
+        expect(res.status).to.be.eql(400);
+        expect(res.body).to.have.property('errors');
+        expect(res.body.errors).to.be.to.a('object');
+        expect(res.body.errors).to.have.property('tags').eql('Only a maximum of 3 tags are allowed');
```
Comment (mazma1) [inline __tests__/articleRoutes.test.js]: Only a maximum of 3 tags are allowed?

Comment (tolulope-od) [reply __tests__/articleRoutes.test.js]: No, this is an outdated commit. It has been increased to 10 now

File: database/models/article.js [hunk 2]
```diff
@@ -27,6 +27,7 @@ module.exports = (sequelize, DataTypes) => {
       },
       ratings: {
         type: DataTypes.INTEGER,
+        defaultValue: 0
```
Comment (mazma1) [inline database/models/article.js:30]: I don't know how you all manage to get away with no trailing commas shaa

Comment (tolulope-od) [reply database/models/article.js:30]: I try to add it as much as I can, but ESLint flags where it's required, I don't get why it's selective about the flagging sometimes

**Comment by mazma1** [comment]
You did a great work on this @tolulope-od, well done!  🎉 🎉

**Pull request merged**